### ISSUE TITLE
source -> inputstream ?

### ISSUE THREAD

**Issue by maximelefrancois86** [sparql-generate/sparql-generate#63]
when doing `SOURCE <> AS ?var` , the whole source is loaded. 

should investigate possibility of the variable being bound to a sub-class of `NodeValue` being a pointer to something an input stream can be created from and open/read/closed one or more times.

this would make 

```
SOURCE <> AS ?var 
ITERATOR iter:XPath( ?var, .... )
```

and 

```
ITERATOR iter:XPath( <>, .... )
```

perform the same

**Comment by aliariff** [comment]
Hi @maximelefrancois86,

We did encounter this issue that lead to out of memory exception. I think we should just try pass a long the value from SOURCE here as the original URI instead of try to open it as input stream and then convert it to string. Let the ITERATOR do the job to get the inputStream as it is the one that really needs the data.

<img width="1001" alt="image" src="https://user-images.githubusercontent.com/6806035/187856521-7bc93b5c-3fcb-4e86-9f33-100bf8b8cd7f.png">

**Comment by maximelefrancois86** [comment]
True. As a workaround some iterators accept a URL as first parameter and open the stream themselves, for example [iter:CSV](http://w3id.org/sparql-generate/iter/CSV). 

What you propose would be the best way to go, but I think it requires extensive work in the library and I don't have the ressources to work on this in the near future.  If you wish to tackle this issue, I'd be happy to approve PRs
